daemon/prompter.c

```diff
@@ -999,6 +1009,56 @@ prompter_cancel_invocation(prompter_t *self)
     }
 }
 
+static void
+prompter_watch_name(prompter_t *self, GDBusConnection *connection, const gchar *name)
+{
+    if( !g_hash_table_contains(self->prm_watchers, name) ) {
+        guint id = g_bus_watch_name_on_connection(
```

Comment:
Looks like also g_bus_watch_name_on_connection() has issues with already gone names. If client exits before daemon has had a change to install a watch, the watch becomes useless. So that e.g.

    for i in $(seq 123); do (dbus-send --type=method_call --system --dest=org.sailfishos.sailjaild1 /org/sailfishos/sailjaild1 org.sailfishos.sailjaild1.PromptLaunchPermissions string:jolla-calendar&); done

(omits --print-reply and thus make dbus-send not wait for reply)

Leads to dialog showing up and not going away despite all clients being gone.

Elsewhere I've used async name owner query after installing name tracker to plug such holes, but I'm not sure if it is worth the effort here (only cosmetic consequences).

Reply:
I did this as another commit on top. It's almost as big change (in lines of code) as the original :grimacing: and it's not even working like it should yet. It actually amplfies the problem by queueing all the dialogs to windowprompt (think of having 123 of the same prompt when you should have just one). :sweat_smile: I think that happens when newPermissionPrompt() has not yet returned but NameHasOwner check already returns. I think we probably should not cancel the call to newPermissionPrompt() but instead take the return value and use the API to cancel it once it returns, however I suspect that gets complicated quite easily. :thinking:

Also that gpointer array is maybe a little ugly, adding a struct with constructor and destructor would be a bit cleaner but feels stupid to add such for so simple case. (Although after I realised I had to dup the string it started to make some sense...) Maybe I should just drop this whole commit as something that doesn't add much value.

Reply:
Making newPermissionPrompt() call non-interruptible was easier by adding two new states and breaking the functions to smaller pieces. This meant fairly significant changes to the states but I think it was worth it. However this means that I doubled the number of changed rows again and you may have to review from the beginning. Sorry. :sweat_smile: 

Since these changes weren't separable and apparently this NameHasOwner check doesn't cause issues anymore, I squashed the changes together.